x/params/store/store.go

```diff
@@ -0,0 +1,167 @@
+package store
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/wire"
+)
+
+// Individual parameter store for each keeper
+type Store struct {
+	cdc  *wire.Codec
+	key  sdk.StoreKey
+	tkey sdk.StoreKey
+
+	space []byte
+}
+
+// NewStore constructs a store with namespace
+func NewStore(cdc *wire.Codec, key sdk.StoreKey, tkey sdk.StoreKey, space string) Store {
+	if !isAlphaNumeric(space) {
+		panic("paramstore space expressions can only contain alphanumeric characters")
+	}
+
+	return Store{
+		cdc:  cdc,
+		key:  key,
+		tkey: tkey,
+
+		space: []byte(space + "/"),
+	}
+}
+
+// Wrapper for key string
+type Key struct {
+	s string
+}
+
+// copied from baseapp/router.go
+var isAlphaNumeric = regexp.MustCompile(`^[a-zA-Z0-9]+$`).MatchString
+
+// Appending two keys with '/' as separator
+// Checks alpanumericity
+func (k Key) Append(keys ...string) (res Key) {
+	res = k
+
+	for _, key := range keys {
+		if !isAlphaNumeric(key) {
+			panic("parameter key expressions can only contain alphanumeric characters")
+		}
+		res.s = res.s + "/" + key
+	}
+	return
+}
+
+// NewKey constructs a key from a list of strings
+func NewKey(keys ...string) (res Key) {
+	if len(keys) < 1 {
+		panic("length of parameter keys must not be zero")
+	}
+	res = Key{keys[0]}
+
+	return res.Append(keys[1:]...)
+}
+
+// KeyBytes make KVStore key bytes from Key
+func (k Key) Bytes() []byte {
+	return []byte(k.s)
+}
+
+// Human readable string
+func (k Key) String() string {
+	return k.s
+}
+
+// Get parameter from store
+func (s Store) Get(ctx sdk.Context, key Key, ptr interface{}) {
+	store := ctx.KVStore(s.key).Prefix(s.space)
+	bz := store.Get(key.Bytes())
+	s.cdc.MustUnmarshalBinary(bz, ptr)
+}
+
+// Get raw bytes of parameter from store
+func (s Store) GetRaw(ctx sdk.Context, key Key) []byte {
+	store := ctx.KVStore(s.key).Prefix(s.space)
+	res := store.Get(key.Bytes())
+	return res
+}
+
+// Check if the parameter is set in the store
+func (s Store) Has(ctx sdk.Context, key Key) bool {
+	store := ctx.KVStore(s.key).Prefix(s.space)
+	return store.Has(key.Bytes())
+}
+
+// Returns true if the parameter is set in the block
+func (s Store) Modified(ctx sdk.Context, key Key) bool {
+	tstore := ctx.KVStore(s.tkey).Prefix(s.space)
+	return tstore.Has(key.Bytes())
+}
+
+// Set parameter, return error if stored parameter has different type from input
+func (s Store) Set(ctx sdk.Context, key Key, param interface{}) error {
+	store := ctx.KVStore(s.key).Prefix(s.space)
+	keybz := key.Bytes()
+
+	bz := store.Get(keybz)
+	if bz != nil {
+		ptrty := reflect.PtrTo(reflect.TypeOf(param))
+		ptr := reflect.New(ptrty).Interface()
+
+		if s.cdc.UnmarshalBinary(bz, ptr) != nil {
+			return fmt.Errorf("Type mismatch with stored param and provided param")
+		}
+	}
+
+	bz, err := s.cdc.MarshalBinary(param)
```

Comment:
* Lets store all parameter values as JSON rather than Binary.  We can figure out a way to cache this in the future (not something we have to do now).